sources/osquery_thrift.js

```diff
@@ -30,6 +38,52 @@ module.exports = class OSQuery {
     }
   }
 
+  static start() {
+    const socket = socketPath[platform] + path.sep + 'shell.em'
+    const binary = osqueryPlatforms[platform]
+    const prefix = process.env.NODE_ENV === 'development' ? '' : '../'
+    const osqueryPath = path.resolve(__dirname, prefix + binary)
+
+    log.info(`startup: ${osqueryPath, ['--nodisable_extensions', `--extensions_socket=${socket}`].join(' ')}`)
+
+    return new Promise((resolve, reject) => {
+      const osqueryi = spawn(osqueryPath, ['--nodisable_extensions', `--extensions_socket=${socket}`], {
+        windowsHide: true,
+      })
+
+      setTimeout(() => {
+        this.connection = ThriftClient.getInstance({ path: socket })
```

Comment:
Can we run some kind of super fast "heartbeat" query at this stage, and retry on failure? Then we're at least polling to see if it's up instead of trying to guess a magic number that will work.

I suppose normal plugins don't have this problem because the load order is reversed: osquery launches *them*.

Reply:
Definitely, that would be a much more ideal setup. I tried something like that at first, but I haven't found any real documentation on the NodeJS API produced by thrift. [This page exists](https://thrift.apache.org/tutorial/nodejs) but all of the links are broken 👎 . I ran this command to generate the thrift lib and have figured out some of the API by reviewing the source:

```bash
thrift -r --gen js:node /path/to/osquery/osquery.thrift
```
I'll dive deeper into the thrift libraries and see what I can do in terms of polling and catching errors.